### DISCUSSION TITLE
Geração do QRCode

### DISCUSSION THREAD

**Discussion by raffamz** [bacen/pix-api#238] [Q&A]
Amigos,
Boa tarde!
Estou implementando a API PIX, tag de **cobrança imediata**. Eu entendia que a criação da cobrança imediata gerava um _qrcode_ e salvava em um domínio ( o mesmo da url do _location_).
Na  leitura da documentação, o psp receber pode gerar o _qrcode_, mas nao nos recursos da cobrança imediata.
Dúvidas:
- temos que gerar o _qrcode_ base64?
- _location_ guarda uma imagem do _qrcode_ ou é um get para um json com dados da cobrança?

Obrigado antecipadamente pela ajuda!

**Comment by rubenskuhl** [comment]
O ideal para ter maior portabilidade é você mesmo gerar o EMV/BRCode e você mesmo gerar a imagem do QR-Code a partir do EMV. O location é JSON com dados de cobrança, que é ligado pelo PSP pagador para entender os parâmetros da cobrança.

**Comment by raffamz** [comment]
Obrigado pela informação.

-O EMV/BRCode é uma necessidade/especificação da API PIX? 
-A url do location é direcionada para o recurso da API PIX CobPayload (/{pixUrlAcessToken})?

Abs

**Comment by rubenskuhl** [comment]
> Obrigado pela informação.
> 
> -O EMV/BRCode é uma necessidade/especificação da API PIX?

É uma necessidade para todos os arranjos de pagamento no Brasil, inclusive o Pix. 
https://www.bcb.gov.br/content/estabilidadefinanceira/spb_docs/ManualBRCode.pdf

> -A url do location é direcionada para o recurso da API PIX CobPayload (/{pixUrlAcessToken})?

Isso mesmo.

**Comment by rubenskuhl** [comment]
Alguns sites que ajudam a montar/desmontar o EMV do Pix:
https://gerarqrcodepix.com.br/
https://pix.ae
https://pix.nascent.com.br/tools/pix-qr-decoder/
https://openpix.com.br/qrcode/scanner/
https://openpix.com.br/qrcode/debug/

**Comment by mauruco** [comment]
Bom dia, amigos.
Tenho uma dúvida de nomenclatura.
{pixUrlAcessToken} não deveria se chamar txid?
Ou {pixUrlAcessToken} é um token temporário que lava a o txid?
Obrigado

**Reply by rubenskuhl** [reply]
> Bom dia, amigos.
> Tenho uma dúvida de nomenclatura.
> {pixUrlAcessToken} não deveria se chamar txid?
> Ou {pixUrlAcessToken} é um token temporário que lava a o txid?
> Obrigado

É um token temporário que leva a várias informações da cobrança, inclusive o txid. Lá tem expiração, recebedor, valor, entre outras.

**Reply by mauruco** [reply]
@rubenskuhl, No criar cobrança o valor no path location seria o {pixUrlAcessToken}, certo? Obrigado pela ajuda. Meio que estou confuso se poderia ser ou não o txid.

**Comment by brazilianbytes** [comment]
Estou integrando, e senti falta de uma lib do Bacen oficial para montar/demonstar esses QRCodes do Pix. Me sentiria mais seguro, já que em serviços abertos ou em libs públicas de terceiros, pode ser algo manipulável. Faz sentido @ninrod?

**Reply by rubenskuhl** [reply]
O BACEN já respondeu negativamente a uma sugestão de incluir endpoints na API Pix para isso, que seria algo bem mais simples do que o BACEN montar libs para n+1 linguagens e seria ônus apenas dos PSPs... quanto mais disponibilizar libs. 

Notar que em não havendo esse recurso padronizado na API Pix, nada impede do seu PSP disponibilizar isso para você... segue um exemplo de um PSP que o fez:
https://dev.gerencianet.com.br/docs/api-pix-endpoints#section-gerar-qrcode-de-um-location

Agora, se você quer algo portável entre PSPs, precisa ou usar libs em que você confie ou escrever a sua a partir das especificações EMV e BRCode.

**Comment by JoseFerreiraThe** [comment]
Gostaria de saber se posso pegar um codigo do qrcodepix em Json e gerar um BMP

**Reply by rubenskuhl** [reply]
Na API padrão não tem o BRCode (Pix Copia e Cola), tem que montar o BRCode primeiro e daí querer o QRCode no formato que você preferir (PNG, JPG, BMP...).

**Reply by JoseFerreiraThe** [reply]
Em meus testes eu gero pela API e abro o QRCode em HTML, mas queria abrir direto em BMP ou JPG

**Reply by rubenskuhl** [reply]
> Em meus testes eu gero pela API e abro o QRCode em HTML, mas queria abrir direto em BMP ou JPG

API de qual prestador ? Pq o padrão do Banco Central não tem isso não...

**Reply by JoseFerreiraThe** [reply]
Usei essa https://github.com/ceciliadeveza/gerarqrcodepix

**Comment by wesleygimenes** [comment]
Bom dia algum código em .net para montar a imagem do Qrcode  a partir do emv?
